### PR TITLE
CNI network plugin

### PR DESCRIPTION
**Pull request by dcbw**
@thockin 

Add an experimental network plugin implementation named "cni" that uses the Container Networking Interface (CNI) specification for configuring networking for pods.

https://github.com/appc/cni/blob/master/SPEC.md

Originally by @rajatchopra

### CODE REVIEW

**Comment by dcbw** [comment]
This PR is a rebase and fixup of https://github.com/kubernetes/kubernetes/pull/13766

**Comment by thockin** [comment]
LGTM - squash and fix nit

**Comment by dcbw** [comment]
@thockin will do

**Comment by dcbw** [comment]
Since I'm a Red Hat employee, I'm covered under the Google Corporate CLA that Red Hat signed on Apr 29, 2015 15:23 PDT (and it is correctly listed at https://cla.developers.google.com/clas while I'm logged in).  So is @rajatchopra ...

**Comment by thockin** [comment]
Are you in the googlegroup you need to be in to indicate your RedHatterness?  @smarterclayton

**Comment by thockin** [comment]
shippable indicated that the cni pkg had a test error - I kicked it just in case, but can you confirm that 'make test' passes on this PR?

**Comment by eparis** [comment]
The commit was authored using @rajatchopra personal gmail address, not his @redhat.com address. `git commit --amend --author="Author Name <email@address.com>"` and then forcepush should make the bot behave better.

**Comment by pmorie** [comment]
@kubernetes/rh-networking

**Comment by dcbw** [comment]
> shippable indicated that the cni pkg had a test error - I kicked it just in case, but can you confirm that 'make test' passes on this PR?

It seems to always fail on Go 1.3, but never on Go 1.4?  I can't get it to fail 'make check' locally with Go 1.4.2 on x86-64.  I'll try to dig into it...

**Comment by dcbw** [comment]
> The commit was authored using @rajatchopra personal gmail address, not his @redhat.com address.

@eparis Fixed that.  How do we kick the CLA bot?

**Comment by thockin** [comment]
I didn't un-lgtm #13766 and it got merged.  can you rebase to get this fixup?

**Comment by bgrant0607** [comment]
I'm going to revert #13766

**Comment by bgrant0607** [comment]
Reverted by #14544

**Comment by thockin** [comment]
That's great, thanks.

**Comment by dcbw** [comment]
@thockin Current status on this one: still chasing the periodic failure; I cannot get it to fail locally with either Go 1.3.3 or Go 1.4.2, but I'm diagnosing the issue.

**Comment by dcbw** [comment]
I signed it!

**Comment by thockin** [comment]
What happened?  The second commit in this PR is obviously wrong..

We're already late for cutoff...

**Comment by dcbw** [comment]
> What happened? The second commit in this PR is obviously wrong..
> 
> We're already late for cutoff...

@thockin Shippable is still failing mostly in the Go 1.3 tests, with "write |1: broken pipe" when executing the test plugin script.  There's no other way to debug into appc/cni than to re-vendor it to myself on github so I can add debugging prints so that Shippable spits out usable information, because the Travis tests verify that the Godeps are the same as upstream, and I cannot change appc/cni to check.

My current theory is that we're hitting https://github.com/golang/go/issues/9173 which is only fixed in Go 1.5, but it's been exceedingly hard to debug because (a) shippable takes forever to complete a test and (b) we can't get any intermediate output from Shippable and (c) we can't easily debug into appc/cni because Go really, really doesn't like changing godeps where the changes are not upstream on github.

**Comment by thockin** [comment]
status?

**Comment by dcbw** [comment]
@thockin I'm testing my theory with a modification to appc that works around the EPIPE thing.  If that passes Shippable I'll submit the PR for appc/cni and hopefully get @eyakubovich to merge it officially, and then I can ditch all my Godeps changes.  Shippable is being quite slow today...

**Comment by dcbw** [comment]
I signed it!

**Comment by thockin** [comment]
status?  Is it worth kicking the jenkins build again?

**Comment by dcbw** [comment]
@thockin it seems to be working in Shippable now at least, I've backed out all the testing commits.  It seems we can work around the issue of the EPIPE in Kubernetes with some magic in the test script.

**Comment by dcbw** [comment]
@thockin I've rebased on top of latest master and added another sanity check to the test script; hopefully this is the end of things.  I think the Jenkins build may have failed due to upstream changes that just got reverted, but I can't actually see Jenkins (I get "access denied").

**Comment by thockin** [comment]
!!! 'gofmt -s' needs to be run on the following files: 
./pkg/kubelet/network/cni/cni_test.go

**Comment by dcbw** [comment]
@thockin Travis and shippable are green now so I think we're good to go.  Suggestions about the CLA thing appreciated; both my and Rajat's addresses are @redhat so we are both covered on the CLA.

**Comment by thockin** [comment]
@k8s-bot test this please

**Comment by thockin** [comment]
I've already poked our CLA folks.

File: pkg/kubelet/dockertools/manager.go [hunk 1]
```diff
@@ -1190,6 +1192,32 @@ func (dm *DockerManager) PortForward(pod *kubecontainer.Pod, port uint16, stream
 	return command.Run()
 }
 
+// Get the IP address of a container's interface using nsenter
+func (dm *DockerManager) GetContainerIP(containerID, interfaceName string) (string, error) {
```
Comment (thockin) [inline pkg/kubelet/dockertools/manager.go]: For future fixup: Could this just be private to the CNI plugin?

Comment (dcbw) [inline pkg/kubelet/dockertools/manager.go]: > For future fixup: Could this just be private to the CNI plugin?

Yeah, it probably could be.  I think Rajat just put in the docker stuff for symmetry, but I think in the future this code will go away when we get the IP from AddNetwork or add a STATUS call to the CNI spec.

**Comment by dcbw** [comment]
@thockin Sorry this one took so long around the release.  I hope my future contributions will have a smoother landing.

**Comment by dcbw** [comment]
I signed it!

**Comment by thockin** [comment]
@k8s-bot test this please

**Comment by thockin** [comment]
@k8s-oncall  The CLA bot is being a PITA - manually confirmed CLA, I'd like to get this merged

**Comment by fabioy** [comment]
Have to improve the CLA thing. Set it to cla:yes and waiting for merge bot.

**Comment by thockin** [comment]
@k8s-bot test this please

**Comment by thockin** [comment]
Lawyers say CLA is OK on this one.

**Comment by thockin** [comment]
I'm manually merging it.

**Comment by jojimt** [comment]
Is there a released package that includes CNI? When I install with source_type == "packageManager", the kubelet does not seem to recognize CNI. If I build the binaries locally, however, it does recognize it.

**Pull request merged**